### PR TITLE
GH-36598: [C++][MinGW] Fix build failure with Protobuf 23.4

### LINKED ISSUE #36598: [C++] CI failures on Windows & Substrait (appears to be abseil linking issues)
### Describe the bug, including details regarding any error messages, version, and platform.

I've seen this CI failure pop up recently on two unrelated PRs.

Example:

https://github.com/apache/arrow/actions/runs/5509303570/jobs/10041817398?pr=35440

```
 FAILED: release/libarrow_substrait.dll release/libarrow_substrait.dll.a 
cmd.exe /C "cd . && D:\a\_temp\msys64\mingw32\bin\c++.exe -Wno-noexcept-type -Wno-self-move  -fdiagnostics-color=always  -Wa,-mbig-obj -Wall -Wno-conversion -Wno-sign-conversion -Wunused-result -Wdate-time -fno-semantic-interposition -mxsave -msse4.2  -O3 -DNDEBUG -O2 -ftree-vectorize  -Wl,--version-script=D:/a/arrow/arrow/cpp/src/arrow/symbols.map -shared -o release\libarrow_substrait.dll -Wl,--out-implib,release\libarrow_substrait.dll.a -Wl,--major-image-version,1300,--minor-image-version,0 @CMakeFiles\arrow_substrait_shared.rsp  && cd ."
D:/a/_temp/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/13.1.0/../../../../i686-w64-mingw32/bin/ld.exe: src/arrow/engine/CMakeFiles/arrow_substrait_shared.dir/substrait/serde.cc.obj:serde.cc:(.text+0xf30): undefined reference to `absl::lts_20230125::operator<<(std::ostream&, absl::lts_20230125::Status const&)'
D:/a/_temp/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/13.1.0/../../../../i686-w64-mingw32/bin/ld.exe: src/arrow/engine/CMakeFiles/arrow_substrait_shared.dir/substrait/serde.cc.obj:serde.cc:(.text+0x101c): undefined reference to `absl::lts_20230125::Status::UnrefNonInlined(unsigned int)'
D:/a/_temp/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/13.1.0/../../../../i686-w64-mingw32/bin/ld.exe: src/arrow/engine/CMakeFiles/arrow_substrait_shared.dir/substrait/serde.cc.obj:serde.cc:(.text+0x141d): undefined reference to `absl::lts_20230125::operator<<(std::ostream&, absl::lts_20230125::Status const&)'
D:/a/_temp/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/13.1.0/../../../../i686-w64-mingw32/bin/ld.exe: src/arrow/engine/CMakeFiles/arrow_substrait_shared.dir/substrait/serde.cc.obj:serde.cc:(.text+0x14f1): undefined reference to `absl::lts_20230125::Status::UnrefNonInlined(unsigned int)'
D:/a/_temp/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/13.1.0/../../../../i686-w64-mingw32/bin/ld.exe: src/arrow/engine/CMakeFiles/arrow_substrait_shared.dir/substrait/serde.cc.obj:serde.cc:(.text$_ZN4absl12lts_202301256StatusD1Ev[__ZN4absl12lts_202301256StatusD1Ev]+0x17): undefined reference to `absl::lts_20230125::Status::UnrefNonInlined(unsigned int)'
D:/a/_temp/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/13.1.0/../../../../i686-w64-mingw32/bin/ld.exe: release/libsubstrait.a(algebra.pb.cc.obj):algebra.pb.cc:(.text$_ZN6google8protobuf13RepeatedFieldIiE14GrowNoAnnotateEii[__ZN6google8protobuf13RepeatedFieldIiE14GrowNoAnnotateEii]+0x8d): undefined reference to `__emutls_v._ZN6google8protobuf8internal15ThreadSafeArena13thread_cache_E'
```

It's very consistent.  I am going to start investigating but wanted to create an issue in case someone was already looking into it.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

There are 2 problems:

* `FindProtobuf.cmake` provided by CMake is incomplete with Protobuf 23.4.
* Misses `-DPROTOBUF_USE_DLLS` for building Substrait related files.

### What changes are included in this PR?

* We need to use `protobuf-config.cmake` provided by Protobuf instead of `FindProtobuf.cmake` provided by CMake because `FindProtobuf.cmake` misses `absl::status` dependency.
* Accept Protobuf 23.4.
* Use `PROTOBUF_USE_DLLS` when we build Substrait related files.
* Use `Boost_INCLUDE_DIRS` instead of `Boost_INCLUDE_DIR` because `Boost_INCLUDE_DIR` isn't defined in `BoostConfig.cmake`.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* Closes: #36598

### CODE REVIEW

**Comment by assignUser** [comment]
Looks like an empty var? https://github.com/apache/arrow/actions/runs/5515207311/jobs/10055287698#step:7:728

**Comment by kou** [comment]
Oh, we needed to use `Boost_INCLUDE_DIRS` (need the last `S`) for portability.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -1711,6 +1711,8 @@ if(ARROW_WITH_PROTOBUF)
     set(ARROW_PROTOBUF_REQUIRED_VERSION "2.6.1")
   endif()
   resolve_dependency(Protobuf
+                     FORCE_ANY_NEWER_VERSION
```
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1714]: Can you add a comment explaning why we add `FORCE_ANY_NEWER_VERSION`?

Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1722]: OK. Done.

**Review by pitrou — APPROVED** [review]
Thanks @kou !

**Pull request merged**